### DISCUSSION TITLE
Schedule not working

### DISCUSSION THREAD

**Discussion by RRoyceus** [aqualinkd/AqualinkD#270] [Q&A]
I have set up scheduling using the App to set the schedule.  Here is the schedule copied from Cron for my VS pump and the cleaner.  

#***** AUTO GENERATED DO NOT EDIT *****
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
0 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=1 -X PUT
2 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=2850 -X PUT
0 8 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=1 -X PUT
0 17 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=0 -X PUT
0 18 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=0 -X PUT
0 19 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=0 -X PUT
#***** AUTO GENERATED DO NOT EDIT *****

It is not working.  Neither pump starts or stops via the schedule.  I have tried using *'s for Day Month, Month, and Day week and the allowed values.  I have also tried leaving the VS pump on at the end of the day and off at the end of the day.

Any help would be appreciated

**Comment by sfeakes** [comment]
Two things to try, first make sure that the command works, second check version of cron.
1) run the below and see if the pump turns on
```curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=1 -X PUT```
If the pump turns on, then it's cron.
2) AqualinkD depends on a version of cron that supports cron.d (not all do), and not knowing your OS I can't say.  if you read the wiki on the scheduler, it will explain more.  ie check file `/etc/defaults/cron` for the existence of `EXTRA_OPTS='-l'` and make sure it's not commented out.
https://github.com/sfeakes/AqualinkD/wiki#Scheduler 

Also look in where ever you system it logging cron to see if their are any errors.

**Reply by RRoyceus** [reply]
Thank you for the quick reply.  The curl command worked.  I am using Debian GNU/Linux 12 (bookworm)
EXTRA OPTS was not commented out.  It reads EXTRA_OPTS='-l'  Your text looks like a -1   Note mine is -l  Could that be it?

There are several options for loging:
# For quick reference, the currently available log levels are:
#   0   no logging (errors are logged regardless)
#   1   log start of jobs
#   2   log end of jobs
#   4   log jobs with exit status != 0
#   8   log the process identifier of child process (in all logs)

 Which should I use?

**Reply by sfeakes** [reply]
The EXTRA_OPTS is -l, (Letter L lower case). On Debian distros this enables reading the files /etc/cron.d directory.

https://manpages.debian.org/bookworm/cron/cron.8.en.html

So it seams either 
A) cron is not running
B) not reading the files in the cron.d directory
C) doesn't like the aqualinkd cron file for some reason.

If you look in your syslog, do you see any cron errors?

You can try modifying aqualinkd cron file and see if the system picks it up.
`sudo touch /etc/cron.d/aqualinkd`

Then you should see a notification in system log about it reading a new file.

`tail -100 /var/log/messages`.  (Messages might be syslog depending on configuration)

Look for something like
`Mar 25 10:03:01 pool cron.info cron[225]: (*system*aqualinkd) RELOAD (/etc/cron.d/aqualinkd)`

You can also look for cron errors with something like
`cat /var/log/messages | grey cron` (again might be syslog and not messages)

Quick search on cron not reading cron.d and some versions of Debian want to see a blank line at the end of the file, aqualinkd does not do that.  So you could try modifying /etc/cron.d/aqualinkd add a blank line to end of file, save, and see if that works.

I you don't see any cron / cron.d messages you may also want to check you have the distro default version of cron installed, Vixie vs some other version that may not support /etc/cron.d

**Comment by RRoyceus** [comment]
Shaun,

Thanks for the help.  I tried putting a space at the end of the file. It did not help, but I am not sure what you meant by blank at the end of the file, i.e. after the comment line or specifically where???

But as seen below, there is a cron file.

***@***.***

But then I run this, it says there is not a file.

sage:  crontab [-u user] [-n] file
            crontab [ -u user ] [ -i ] { -e | -l | -r }

            -h        (displays this help message)

            file       (default operation is replace, per 1003.2)
            -n        (dry run: checks the syntax, then bails out)
            -u user           (choose the user whose crontab is touched)

            -e         (edit user's crontab)
            -l          (list user's crontab)
            -r          (delete user's crontab)

            -i          (prompt before deleting user's crontab)


***@***.***:~ $ crontab -l
no crontab for pi

Best Regards,

Jimmy Scruggs

The Scruggs Company
8530 Hansen Rd
Houston, TX 77075

Cell : 713-899-1862
Office: 713-649-2776
***@***.***

**Comment by sfeakes** [comment]
Sorry, I meant a blank line at the end of the file.

Your current /etc/conf.d/aqualinkd looks like below. (ignoring `<start>` and `<end>`
```
<start>
#***** AUTO GENERATED DO NOT EDIT *****
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
0 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=1 -X PUT
2 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=2850 -X PUT
0 8 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=1 -X PUT
0 17 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=0 -X PUT
0 18 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=0 -X PUT
0 19 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=0 -X PUT
#***** AUTO GENERATED DO NOT EDIT *****
<end>
```

Change it to
```
<start>
#***** AUTO GENERATED DO NOT EDIT *****
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
0 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=1 -X PUT
2 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=2850 -X PUT
0 8 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=1 -X PUT
0 17 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=0 -X PUT
0 18 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=0 -X PUT
0 19 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=0 -X PUT
#***** AUTO GENERATED DO NOT EDIT *****

<end>
```
(Blank line at end but doesn't show well in this page, so I added `<start>` & `<end>`

The crontab command will only shows information in `/var/spool/cron/crontabs/`, not application cron information in `/etc/cron.d/` so you can ignore using that to debug this.

Few other things to look at.

To check the version of cron installed on your system, can you look at the first few lines output after running `man cron`, hopefully you see
```
NAME
       cron - daemon to execute scheduled commands (Vixie Cron)
```
`Vixie` is important here, as that's the default cron for Debian. (but their are many others that can be installed)
If it's not Vixie can you let me know.

Can you also post the output of the below command, let's see if there is anything logged.
```
`cat /var/log/syslog | grep cron`
```
If that fails or is blank try
```
`cat /var/log/messages | grep cron`
```

can you also post the output of the below command.
```
cat /etc/os-release
```

**Comment by RRoyceus** [comment]
Shaun

Thanks again.  I inserted the blank line.

My Cron:
NAME
       cron - daemon to execute scheduled commands (Vixie Cron)

Using your commands:
***@***.***:~ $ cat /var/log/syslog | grep cron
cat: /var/log/syslog: No such file or directory
***@***.***:~ $ cat /var/log/messages | grep cron
cat: /var/log/messages: No such file or directory

I will see if my schedules start in the morning.

Best Regards,

Jimmy Scruggs

713-899-1862
***@***.***

**Comment by RRoyceus** [comment]
Shaun,

I forgot to send you this:
***@***.***:~ $ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL=https://www.debian.org/
SUPPORT_URL=https://www.debian.org/support
BUG_REPORT_URL=https://bugs.debian.org/

Best Regards,

Jimmy Scruggs

713-899-1862
***@***.***

**Comment by sfeakes** [comment]
I just did a fresh install of `Debian GNU/Linux 12 (bookworm)` the exact release from your os-release and these are the steps I had to do to make it work.

Edit file `/etc/default/cron` and remove comment from line `EXTRA_OPTS='-l'` ie remove `#` at start of line.
restart cron `systemctl restart cron`

Then everything worked as expected.

I did notice that Debian 12 has moved over completely to journal, so that's why the commands I told you to use for looking at the logs didn't work.   ie it uses journal for all system logging.  So to look at the logs you can use the below commands. (you should be root to see everything).
`journalctl` - will list everything
`journalctl | grep cron` - will list cron messages
`journalctl --follow` - will print logs in realtime (well not realtime, a few seconds late due to buffering).

So use `journalctl | grep cron` and see if you see anything about aqualinkd or cron.d.

**Comment by RRoyceus** [comment]
Shaun,

Once again, your email says '-l', but I assume you mean by that lower case L.  When I type, my lowercase L also shows it as 1.  I am using lower case L.

The log took several minutes to list.  I remember a way to limit the size of the log, but I can’t remember what it was.  Here are the last few lines:

Mar 26 18:46:07 raspberrypi cron[13827]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)
Mar 26 18:46:07 raspberrypi cron[13827]: (CRON) INFO (Skipping @reboot jobs -- not system startup)
Mar 26 18:47:01 raspberrypi cron[13827]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)
Mar 26 18:47:41 raspberrypi polkitd[606]: Operator of unix-session:1 successfully authenticated as unix-user:pi to gain TEMPORARY authorization for action org.freedesktop.systemd1.manage-units for system-bus-name::1.66 [systemctl restart cron] (owned by unix-user:pi)
Mar 26 18:47:41 raspberrypi systemd[1]: Stopping cron.service - Regular background program processing daemon...
Mar 26 18:47:41 raspberrypi systemd[1]: cron.service: Deactivated successfully.
Mar 26 18:47:41 raspberrypi systemd[1]: Stopped cron.service - Regular background program processing daemon.
Mar 26 18:47:41 raspberrypi systemd[1]: Started cron.service - Regular background program processing daemon.
Mar 26 18:47:41 raspberrypi cron[13861]: (CRON) INFO (pidfile fd = 3)
Mar 26 18:47:41 raspberrypi cron[13861]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)
Mar 26 18:47:41 raspberrypi cron[13861]: (CRON) INFO (Skipping @reboot jobs -- not system startup)
Mar 26 18:48:01 raspberrypi cron[13861]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)
Mar 26 18:49:01 raspberrypi cron[13861]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)
Mar 26 18:50:01 raspberrypi cron[13861]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)
Mar 26 18:51:01 raspberrypi cron[13861]: (*system*aqualinkd) INSECURE MODE (group/other writable) (/etc/cron.d/aqualinkd)sud

Here is the cron:
#***** AUTO GENERATED DO NOT EDIT *****
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
50 18 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=1 -X PUT
1 7 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/RPM/set -d value=2350 -X PUT
0 8 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=1 -X PUT
0 17 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Aux_1/set -d value=0 -X PUT
0 18 * * * root curl -s -S --show-error -o /dev/null localhost:80/api/Filter_Pump/set -d value=0 -X PUT
#***** AUTO GENERATED DO NOT EDIT *****

I also have a blank line at the end after the last comment.  Notice, I had changed it to turn on the pump at 6:50 (18:50) tonight. It did not turn on.

I also notice in /etc/cron.d/ there is a file “e2scrub_all”  that has this, if it means anything to you:
30 3 * * 0 root test -e /run/systemd/system || SERVICE_MODE=1 /usr/lib/aarch64-linux-gnu/e2fsprogs/e2scrub_all_cron
10 3 * * * root test -e /run/systemd/system || SERVICE_MODE=1 /sbin/e2scrub_all -A -r

Maybe it is time to do a fresh install of bookworm and the program???

Best Regards,

Jimmy Scruggs

713-899-1862
***@***.***

**Comment by sfeakes** [comment]
I don’t think a fresh install will do anything.  I can see from the first log that it’s picked up the AqualinkD cron file, but doesn’t like the permissions of it.
can you try this
chmod 600 /etc/cron.d/aqualinkd 

Then see if it works.   Look at the journal log for cron saying it read aqualinkd and not complaining about insecure mode.

**Reply by sfeakes** [reply]
Forgot probably best to restart cron after you change the permission of the file

**Comment by RRoyceus** [comment]
It worked!!!!  You are the man!  Much thanks for the fast replies and sticking with me on this.

I had to use sudo:
***@***.***:~ $ chmod 600 /etc/cron.d/aqualinkd
chmod: changing permissions of '/etc/cron.d/aqualinkd': Operation not permitted
***@***.***:~ $ sudo chmod 600 /etc/cron.d/aqualinkd
***@***.***:~ $ systemctl restart cron
***@***.***:~ $

It looks like the log is getting way to big.  What can I do to limit the size?  In the mean time, where do I find the file to erase most of it.  Should I just use sudo nano?

Best Regards,

Jimmy Scruggs

713-899-1862
***@***.***

**Comment by sfeakes** [comment]
Glad we got to the bottom of this,  I’ll modify the code to make sure this doesn’t happen to anyone else.

journal should keep its size down without depending on other tools to do it, it’s one of the advantages it has over syslog. So you probably don’t have to worry. But Here is a quick guide that may help you https://www.loggly.com/ultimate-guide/managing-journal-size/
